### ISSUE TITLE
uninitialized constant LogStash::Outputs::Tcp::TCPSocket

### ISSUE THREAD

**Issue by Gingonic** [elastic/logstash#3036]
Hi

while trying to output to a TCP socket, Logstash throws following error

```
tcp output exception {:host=>"XXXXXXXX", :port=>31051, :exception=>#<NameError: uninitialized constant LogStash::Outputs::Tcp::TCPSocket>, :backtrace=>["org/jruby/RubyModule.java:2690:in `const_missing'", 
"/opt/logstash/lib/logstash/outputs/tcp.rb:124:in `connect'", "org/jruby/RubyProc.java:271:in `call'",
"/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/try.rb:75:in `try'", 
"org/jruby/RubyProc.java:271:in `call'", 
"/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/try.rb:91:in `try'", 
"/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/try.rb:20:in `each'", 
"/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/try.rb:87:in `try'", 
"/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/try.rb:119:in `try'", 
"/opt/logstash/lib/logstash/outputs/tcp.rb:123:in `connect'", 
"/opt/logstash/lib/logstash/outputs/tcp.rb:100:in `register'", "org/jruby/RubyProc.java:271:in `call'", 
"/opt/logstash/lib/logstash/codecs/json.rb:45:in `encode'", 
"/opt/logstash/lib/logstash/outputs/tcp.rb:143:in `receive'", 
"/opt/logstash/lib/logstash/outputs/base.rb:86:in `handle'", "(eval):17:in `initialize'", 
"org/jruby/RubyProc.java:271:in `call'", "/opt/logstash/lib/logstash/pipeline.rb:266:in `output'", 
"/opt/logstash/lib/logstash/pipeline.rb:225:in `outputworker'", 
"/opt/logstash/lib/logstash/pipeline.rb:152:in `start_outputs'"], :level=>:warn}
```
I tested the socket, messages can pass through

``` 
nc XXXXXXXX 31051
```  
I'have installed logstash 1.4.2 over ubuntu 14.04 package 

The input consumer works and I don't think the input can have an impact but still : 

```
input {
  rabbitmq {
    queue => "log-messages"
    host => "YYYYYYY"
    password => "somepass"
    port => 31075
    user => "someuser"
    exchange => "log-messages"
    key => "log-key"
  }
}
output {
  tcp {
    host => "XXXXXXXX"
    port => "31051"
  }
}
```

Any idea ? I could only find one similar issue back in february 2014 on the logstash 1.3, but no answer.

**Comment by wiibaa** [comment]
it seems that tcp output is missing a line `require "socket"`, so in your minimal it is not loaded by another plugin and then failed, you could try to add it either at the top of the file after the other require statements

**Comment by purbon** [comment]
I'm adding test to the solution proposed by @wiibaa so we're sure the plugin works as expected :-)

**Comment by purbon** [comment]
@Gingonic I did you very same test with the 1.5 branch and this problem was not shown to me. Can you test with 1.5 and confirm you don't have this issue ?

I test also with a 1.4.2 branch (from the tgz file) and I can not see this very same issue. I will check also with the 1.4.2 package, and report back here. Can you share more details of what are you doing? so we can reproduce more clearly.

In my case I did a basic test with:

```
83.149.9.216 - - [24/Feb/2015:23:13:42 +0000] "GET /presentations/logstash-monitorama-2013/images/kibana-search.png HTTP/1.1" 200 203023 "http://semicomplete.com/presentations/logstash-monitorama-2013/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1700.77 Safari/537.36"
83.149.9.216 - - [24/Feb/2015:23:13:42 +0000] "GET /presentations/logstash-monitorama-2013/images/kibana-dashboard3.png HTTP/1.1" 200 171717 "http://semicomplete.com/presentations/logstash-monitorama-2013/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1700.77 Safari/537.36"
83.149.9.216 - - [24/Feb/2015:23:13:44 +0000] "GET /presentations/logstash-monitorama-2013/plugin/highlight/highlight.js HTTP/1.1" 200 26185 "http://semicomplete.com/presentations/logstash-monitorama-2013/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1700.77 Safari/537.36"
```

```
╰─$ cat ../test.conf 
input {
  stdin {}
}
output {
  tcp {
    host => "127.0.0.1"
    port => "31051"
  }
}
```
and the noted versions.

**Comment by wiibaa** [comment]
@purbon the point is that even stdin is doing `require "socket" # for Socket.gethostname` but not the rabbitmq input

**Comment by purbon** [comment]
Ok.
   i really assumed stdin was not messing with sockets :-), thanks for the
tip.

/purbon

On Friday, April 17, 2015, Philippe Weber <notifications@github.com> wrote:

> @purbon <https://github.com/purbon> the point is that even stdin is doing require
> "socket" # for Socket.gethostname but not the rabbitmq input
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/logstash/issues/3036#issuecomment-94031858>.
>


-- 
Pere Urbon-Bayes
Software Engineer
https://www.elastic.co
https://twitter.com/purbon

**Comment by colinsurprenant** [comment]
for the sake of this test you could use a "noop" plugin similar to what we do in spec/filters/base_spec.rb that way you can make sure nothing is included.

**Comment by purbon** [comment]
@wiibaa you fix proposal makes complete sense to me, I'm just wondering a couple of things:

* Should socket be required before and globalish as it's so common and popular among the plugins?
* A way to rest this missing and exception in an rspec test suite as the plugin has really no test at all.

**Comment by purbon** [comment]
@colinsurprenant this is a good idea, I think we should have a set of noops very simple raw input/filter/output very good for debugging.

**Comment by purbon** [comment]
@wiibaa @Gingonic 

I did a very simple test this morning trying to reproduce this error by using a noop input and the tcp output.

The input plugin I used

```
# encoding: utf-8
require "logstash/inputs/base"
require "logstash/namespace"

# A simple noop input plugin.
class LogStash::Inputs::Noop < LogStash::Inputs::Base
  config_name "noop"

  # If undefined, Logstash will complain, even if codec is unused.
  default :codec, "plain"

  public
  def register
  end # def register

  def run(queue)
      event = LogStash::Event.new("message" => "Hello World!")
      decorate(event)
      queue << event
  end # def run

end # class LogStash::Inputs::Noop
```
and the conf

```
╰─$ cat ../test.conf                                                                                                                                       1 ↵
input {
  noop {
  }
}
output {
  tcp {
    host => "127.0.0.1"
    port => "31051"
  }
}
```

while I see that your fix is reasonable @wiibaa, I'm still unable to isolate and reproduce this problem.

**Comment by wiibaa** [comment]
@purbon I'm also having a bad time thinking of a testcase in 1.5 because of https://github.com/elastic/logstash/blob/master/lib/logstash/patches/bugfix_jruby_2558.rb#L4
are you also testing on windows currently ?

**Comment by purbon** [comment]
@wiibaa nope, doing linux and macosx.

**Comment by purbon** [comment]
In my experience usually an error like this

```uninitialized constant LogStash::Outputs::Tcp::TCPSocket```

means a code problem, actually something like what was happening in  https://github.com/elastic/logstash/blob/master/lib/logstash/pluginmanager/command.rb#L36 and the way to fix it is throw appending ```::``` before the name, so I'm suspecting he has an old version that might has this bug, however the actuall code base has not the same problem.

**Comment by Gingonic** [comment]
I'm on the "1.4.2-modified" version. I'm using logstash to simply parse a big legacy log file and push the results through tcp to graylog2. I used RabitMQ as a buffer in between. So my config file is pretty much looking like I posted above. The proposed fix of @wiibaa did solve the problem for me.

**Comment by wiibaa** [comment]
@Gingonic thanks that it has unlocked your situation

@purbon it is easy to reproduce in 1.4.2 with your config, but impossible in 1.5.0rc2 because something is already loading the socket library. This I identified by comparing the output obtained with logstash/patches/profile_require_calls in 1.4.2 and 1.5. In conclusion the main difference is that runner.rb now contains `LogStash::Environment.bundler_setup!` that does a `require 'bundler'` that need net/http that need socket and voila!

**Comment by purbon** [comment]
ok! 
   so as we're all set in here, @Gingonic seems to have a fix for his situation in 1.4.2 and we identified where the socket lib has been loaded in 1.5. I will open/merge the related issues so we get this fix. 

Thanks @wiibaa and @Gingonic for reporting and proposing a fix for this.
